Submit tests for intellisense (replacing functional tests we had)

rchiodo opened the pull request:
For #7028 

I worked on this a week ago and was waiting on the python extension to submit changes I made. They should be there in the python insiders bits.

However I also want these tests in place for the work I'm doing to move intellisense to the jupyter extension. Once these are in, I'll make changes/add to them to test the work I'm doing to create a language server per interpreter.

rchiodo committed:
Tests working independently

rchiodo committed:
Use dev build of python extension

rchiodo committed:
Fix build error

rchiodo committed:
Do it without azure login

rchiodo committed:
Add some more logging

rchiodo committed:
Add more logging for fetching completions

rchiodo committed:
Add more logging and wait some in the test

rchiodo committed:
Update logging

rchiodo committed:
Fix linter

rchiodo committed:
Dont use old jedi

rchiodo committed:
See if csv is actually loading

rchiodo committed:
Print out current folder during test

rchiodo committed:
Cell output not in string form

DonJayamanne commented on src/client/datascience/notebook/intellisense/completionProvider.ts:
```diff
@@ -76,7 +77,10 @@ export class NotebookCompletionProvider implements CompletionItemProvider {
+            traceInfoIf(isCI, `Notebook completions not found.`);
             return [];
+        } else {
+            traceInfoIf(isCI, `Completions found, filtering the list: ${JSON.stringify(result)}.`);
```
I think we should just turn this into `traceIfCI`, lol, feels kinda redundant to always pass in `isCI`. What do you think.
We might have a few places where we'd have some other conditions being passed in, but i think traceIfCI` or `traceOnCI` is simpler.
Absolutely no need to make that change, just a thought

rchiodo replied:
Yeah sounds like a good idea. I'll add a debt issue

DonJayamanne commented on src/test/datascience/notebook/intellisense/diagnostics.vscode.test.ts:
```diff
@@ -0,0 +1,76 @@
+        assert.ok(diagnostics.find((item) => item.message.includes('system')));
+    });
+    test('Markdown cells dont get errors', async () => {
+        // Create two cells, one markdown, make sure not diagnostics in markdown
```
Nice.

rchiodo committed:
Fix read csv code to work with default notebookFileRoot

rchiodo committed:
Add screenshot and some more logging (looks like execution isn't working)

rchiodo committed:
Actually capture a screenshot

rchiodo committed:
Skip another test and make sure selected controller is accurate

rchiodo committed:
Fix compile errors

rchiodo committed:
Add more logging and try turning off python extension experiments a different way

rchiodo committed:
Add verbose logging to VS code

rchiodo committed:
Add VS code logs

rchiodo committed:
Change the way waitForKernelAutoSelect works. Have it loop.

rchiodo committed:
Update timeout for affinity. Can't wait too long

rchiodo committed:
Fix non python cases

rchiodo committed:
Refactor how selection changes occur

rchiodo committed:
Fix some test problems and more logging

rchiodo committed:
Two more fixes

rchiodo committed:
Merge remote-tracking branch 'origin/main' into rchiodo/intellisense_tests

rchiodo committed:
Disable some more tests that actually found a bug. Update variable view to use waitForCondition instead of trying to track messages

rchiodo commented:
@IanMatthewHuff and @DonJayamanne can you take a second look? 

I reworked how we select a kernel in the tests. It now loops on the notebook.selectKernel function. But the default it picks is lot more complicated now. Actually found a bug with this change: https://github.com/microsoft/vscode-jupyter/issues/7610

I also reworked how variables are waited for in tests. Instead of guessing the messages to be sent, I switched it to just use ```waitForCondition``` which should work regardless of how many messages get passed around.

IanMatthewHuff commented on src/client/datascience/notebook/vscodeNotebookController.ts:
```diff
@@ -153,10 +157,10 @@ export class VSCodeNotebookController implements Disposable {
-            await this.setAsActiveControllerForTests(notebook);
+
+        // Keep track of if this controller is preferred for this notebook
+        if (affinity === NotebookControllerAffinity.Preferred) {
```
updateNotebookAffinity is called on each document open if you do something like this:
Open foo.ipynb (A kernel in metadata) foo.ipynb added to A perferredWith
Change to kernel B in foo.ipynb
Execute
Close foo.ipynb
Open foo.ipynb (B kernel in metadata) foo.ipynb added to B preferredWith

Seems like that situation would be wrong then if I'm following right.

rchiodo replied:
Not following the mistake? Because A would still be preferred? Yeah seems like map should be the other way. Map between notebook (key) and controller (value)

IanMatthewHuff replied:
Yeah, trying to make sure that I'm reading it right. As I see it now PreferredWith is all the notebook documents that this controller is the preferred controller for. Given that the preferred controller could change if you execute and reopen a document you could have the same document in the preferred list for two different controllers. As documents only have one preferred (at least from our extensions perspective) that situation wouldn't be right.

IanMatthewHuff commented on src/test/datascience/variableView/variableView.vscode.test.ts:
```diff
@@ -102,17 +92,12 @@ suite('DataScience - VariableView', function () {
             { name: 'test2', type: 'str', length: '12', value: ' MYTESTVALUE2' }
         ];
-        verifyViewVariables(expectedVariables, htmlResult);
+        await waitForVariablesToMatch(expectedVariables, variableView);
```
Nice, much better.

rchiodo committed:
Rework how preferred is cached

rchiodo committed:
Switch back to stable python

rchiodo committed:
Fix linter

rchiodo committed:
Comment out intellisense tests for now

rchiodo committed:
Rework how notebook controller is found for the interactive window. Test was not finding it even though it was using it to run

rchiodo committed:
disable some more tests and try to see why interactive test is failing

DonJayamanne commented on .github/workflows/build-test.yml:
```diff
@@ -43,6 +43,8 @@ env:
   DISABLE_INSIDERS_EXTENSION: 1 # Disable prompts to install Insiders in tests (else it blocks activation of extension).
   VSC_JUPYTER_INSTRUMENT_CODE_FOR_COVERAGE: true
+  VSIX_NAME_PYTHON: 'ms-python-insiders.vsix'
+  VSC_JUPTYER_PYTHON_EXTENSION_VERSION: 'stable'
```
Does this mean, if this is stable, then we use Stable VS Code & stable Python extension
& if its insider we use vscode insiders & python insiders?

rchiodo replied:
No it's just a flag for now. It makes it easy to switch this later.

rchiodo committed:
Add screenshot to test failure

rchiodo committed:
Dont check env name, may not be there

DonJayamanne commented on src/client/datascience/interactive-window/interactiveWindow.ts:
```diff
@@ -629,7 +625,7 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
             { preserveFocus: true },
             notebookDocument.uri,
-            this.notebookController?.id,
+            this.notebookControllerManager.getSelectedNotebookController(notebookDocument)?.id,
```
I'm assuming these changes can be revereted now?
Would actually prefer to go back to the older API (if we can confirm the issues were with the Python extension)

rchiodo replied:
The older api was saving the controller. This queries for it at run time. Seems more robust this way to me.

DonJayamanne replied:
Aaah yes, users can switch the kernels, thats right.

DonJayamanne commented on src/client/datascience/notebook/vscodeNotebookController.ts:
```diff
@@ -408,31 +404,4 @@ export class VSCodeNotebookController implements Disposable {
-     * This method is called on when running tests, else in the real world,
-     * users need to select a kernel (preferred is on top of the list).
-     */
-    private async setAsActiveControllerForTests(notebook: NotebookDocument) {
```
Not sure why thi was deleted, was this to do with getting tets to pass?

rchiodo replied:
Because it doesn't work (notebook.selectKernel has to be called in a loop) and it masks other problems (it always associates the controller with the notebook even when it doesn't work).

I moved all of this same logic into the test helper.ts which does select in a loop.

rchiodo replied:
The new way seems more likely to find issues as it forces the UI selection change before the controller actually changes.

DonJayamanne replied:
Dont' think tests will pass without this, as without this users will be prompted to select a kernel (this ensures the kernel picker isnt displayed in tests.)

rchiodo replied:
No it does the same thing it used to do. It sends the command 'notebook.selectKernel' in a loop. This always happens in all tests before runnign a cell. The old code used to call this setAsActiveControllerForTests instead. Now it just waits for the UI to work.

rchiodo replied:
Meaning the selection to change (not for the select kernel picker to come up). 'notebook.selectKernel' does not put up the picker if you pass an id.

DonJayamanne replied:
Yup, found it. All good

DonJayamanne commented on src/client/datascience/notebook/types.ts:
```diff
@@ -13,6 +13,7 @@ export interface INotebookControllerManager {
     // Marked test only, just for tests to access registered controllers
     registeredNotebookControllers(): VSCodeNotebookController[];
     getInteractiveController(): Promise<VSCodeNotebookController | undefined>;
+    getPreferredNotebookController(document: NotebookDocument): VSCodeNotebookController | undefined;
```
I don't think this is being used.

rchiodo replied:
It's used in the helper.ts file to find the closest match for a kernel.

rchiodo replied:
Basically the the tests that used to call waitForKernelToGetAutoSelected are now instead waiting for the preferred kernel to be auto selected (which effectively is the same as what happened before)

DonJayamanne replied:
Yes found that.

DonJayamanne commented on src/test/datascience/notebook/helper.ts:
```diff
@@ -287,89 +305,88 @@ export async function waitForKernelToChange(criteria: { labelOrId?: string; inte
+                if (!isRightKernel()) {
+                    traceInfoIf(isCI, `Notebook select.kernel command switching to kernel id ${id}: Try ${tryCount}`);
+                    // Send a select kernel on the active notebook editor. Keep sending it if it fails.
+                    await commands.executeCommand('notebook.selectKernel', { id, extension: JVSC_EXTENSION_ID });
```
I dont think this needs to be ezecuted multiple times, i think once is sufficient, and after that we just need to keep waiting until the right kernel is selected.
Else we keep firing this command multkiple times & i'm concerned we could have other side effects of filing this unnecessarily

rchiodo replied:
That's not what I found. There was a potential race between 'createNotebookController (and results getting to VS code proper)' and the 'notebook.selectKernel' command coming in. This race went away when I did this in a loop.

rchiodo replied:
Previously we didn't even care if it worked (hence the comment in the ```setAsActiveControllerForTests```). And we'd set and forget it. But sometimes it didn't work.

With this loop it seems to always work.

DonJayamanne reviewed (approved):
Looks good
